compiler-rt/lib/scudo/standalone/secondary.h

```diff
@@ -351,9 +351,13 @@ class MapAllocatorCache {
     for (MemMapT &EvictMemMap : EvictionMemMaps)
       unmapCallBack(EvictMemMap);
 
-    if (Interval >= 0) {
+    // If a thread already holds the mutex, the current thread can
```

Comment:
I don't think this is strictly true, right? Specifically, in the retrieve function you could hold the lock while trying to find an entry in the cache.

Is there a worry that you could skip these checks while doing a retrieve?

Reply:
Right, I misunderstood why this was originally added, but I believe that in the case that many retrieves are happening and blocking the acquisition of the lock, we decided it was okay to skip the release to avoid releasing memory right before it was retrieved anyway. I will update the comment accordingly.

Reply:
I've updated the comment, but now that I think about it, do you think it would be worth adding a lock specifically for releases? So instead we would have something like 

```
if (Interval >= 0 && ReleaseMutex.tryLock()) {
      releaseOlderThan(Time);
      ReleaseMutex.unlock();
}
```
where `releaseOlderThan()` acquires the cache lock inside the function?

Reply:
Sorry I didn't think about this carefully and I think @cferris1000 is right.

When a page releasing is skipped, another store() or retrieve() can be the reason (there are some other situations which needs the lock too but I think they are rare)

1. When another store() happens, it will try the page releasing later. So this case is fine.

2. When it's skipped by retrieve(), we may miss the chance to release a bunch of pages when they all pass the release-to-os interval. For example, assuming the release-to-os interval is 1s. Suppose we have filled out all the entries within 1s. After another 1s, there are a retrieve() and a store() happen at the same time. If we miss the page releasing in the store(), that means we leave almost all the entries (except the one retrieved) unreleased. This can add some pressure to the memory usage.

This may or may not be a real problem because the case may be rare but it does have the chance to increase memory usage. Let's jump back to the motivation of this change, do we see a lot of contention here?

BTW, adding another lock doesn't work because the entries are guarded by the `Mutex`.

Reply:
I ran some checks internally, and it seems that even on processes with the most threads (~200), there is only about 1 or 2 contended acquisitions of the lock per 1000 calls to store, so it's probably not justified to add this if we're concerned about the second case.

Just to clarify, however, I wasn't saying we should replace `Mutex` but rather add another `HybridMutex` if we wanted to avoid the second case. It would look something like
```
if (Interval >= 0 && ReleaseMutex.tryLock()) {
      CacheMutex.lock()
      releaseOlderThan(Time);
      CacheMutex.unlock();
      ReleaseMutex.unlock();
}
```
where `CacheMutex` is the original `Mutex` in the current code. However, given that there doesn't seem to be much of a need for reducing contention here, I think we can close this pull request for now. Let me know what you think.